src/beamlime/constructors/injection.py

```diff
@@ -0,0 +1,171 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+from __future__ import annotations
+
+from typing import Any, Dict, Type
+
+from .base import (
+    FactoryBase,
+    MismatchingProductTypeError,
+    ProviderExistsError,
+    ProviderNotFoundError,
+    _Product,
+)
+from .inspectors import DependencySpec, ProductType, UnknownType
+from .providers import Constructor, Provider
+
+
+class RegistrationInterface(FactoryBase):
+    """Provider registration interfaces."""
+
+    def register(self, product_type: ProductType, provider_call: Constructor) -> None:
+        """
+        Register a provider function ``provider_call`` of ``product_type``
+        into the ``self._providers`` dictionary.
+
+        If ``product_type`` is ``UnknownType``,
+        it will retrieve the product type from the provider call
+        and use it as an index.
```

Comment:
Can you expand the docs to explain when it is possible to retrieve the type and how to use this? An example may help.